File: days/day_12/solutions/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

void solve(char* content, char** part1_result, char** part2_result) {
    // Split content into lines
    char** lines = NULL;
    int line_count = 0;
    int capacity = 10000;
    lines = (char**)malloc(capacity * sizeof(char*));
    
    char* pos = content;
    while (*pos != '\0') {
        char* line_start = pos;
        // Find next newline or end of string
        while (*pos != '\n' && *pos != '\0') {
            pos++;
        }
        
        if (line_count >= capacity) {
            capacity *= 2;
            lines = (char**)realloc(lines, capacity * sizeof(char*));
        }
        
        if (*pos == '\n') {
            *pos = '\0';
            pos++;
        }
        
        lines[line_count++] = line_start;
    }
    
    if (line_count == 0) {
        *part1_result = strdup("0");
        *part2_result = strdup("Final star");
        free(lines);
        return;
    }
    
    // Parse shapes (first 6 shapes, numbered 0-5)
    int shape_areas[6] = {0, 0, 0, 0, 0, 0};
    int i = 0;
    int shape_idx = 0;
    
    while (i < line_count && shape_idx < 6) {
        // Trim line
        char* line = lines[i];
        while (*line == ' ' || *line == '\t') line++;
        int line_len = strlen(line);
        while (line_len > 0 && (line[line_len-1] == ' ' || line[line_len-1] == '\t' || line[line_len-1] == '\r')) {
            line[line_len-1] = '\0';
            line_len--;
        }
        
        // Check if this is a shape header (format: "number:")
        if (line_len > 0 && line[line_len-1] == ':') {
            line[line_len-1] = '\0';
            int shape_num = -1;
            if (sscanf(line, "%d", &shape_num) == 1 && shape_num == shape_idx) {
                // Read the next 3 lines for the shape grid
                int area = 0;
                for (int j = 0; j < 3; j++) {
                    if (i + 1 + j < line_count) {
                        char* row = lines[i + 1 + j];
                        // Count '#' characters
                        for (int k = 0; row[k] != '\0'; k++) {
                            if (row[k] == '#') {
                                area++;
                            }
                        }
                    }
                }
                shape_areas[shape_idx] = area;
                shape_idx++;
                i += 4; // Skip shape header + 3 grid lines + empty line (if present)
                continue;
            }
        }
        i++;
    }
    
    // Find where queries start (skip empty lines after shapes)
    int query_start = i;
    while (query_start < line_count) {
        char* line = lines[query_start];
        while (*line == ' ' || *line == '\t') line++;
        if (*line != '\0') break;
        query_start++;
    }
    
    // Parse queries
    int possible_count = 0;
    for (int line_idx = query_start; line_idx < line_count; line_idx++) {
        char* line = lines[line_idx];
        // Trim
        while (*line == ' ' || *line == '\t') line++;
        int line_len = strlen(line);
        while (line_len > 0 && (line[line_len-1] == ' ' || line[line_len-1] == '\t' || line[line_len-1] == '\r')) {
            line[line_len-1] = '\0';
            line_len--;
        }
        if (line_len == 0) continue;
        
        // Parse query: "widthxheight: count0 count1 count2 count3 count4 count5"
        char* colon = strchr(line, ':');
        if (!colon) continue;
        
        *colon = '\0';
        char* dims = line;
        char* counts_str = colon + 1;
        
        // Parse dimensions
        char* x = strchr(dims, 'x');
        if (!x) {
            *colon = ':';
            continue;
        }
        
        *x = '\0';
        int width = atoi(dims);
        int height = atoi(x + 1);
        
        // Parse counts
        int counts[6];
        int count_idx = 0;
        char* token = strtok(counts_str, " \t");
        int valid = 1;
        while (token && count_idx < 6) {
            counts[count_idx++] = atoi(token);
            token = strtok(NULL, " \t");
        }
        
        if (count_idx != 6) {
            *x = 'x';
            *colon = ':';
            continue;
        }
        
        // Calculate area check
        long long region_area = (long long)width * (long long)height;
        long long required_area = 0;
        for (int j = 0; j < 6; j++) {
            required_area += (long long)shape_areas[j] * (long long)counts[j];
        }
        
        if (required_area <= region_area) {
            possible_count++;
        }
        
        *x = 'x';
        *colon = ':';
    }
    
    // Part 2: Final star (no computation needed)
    *part1_result = (char*)malloc(32);
    snprintf(*part1_result, 32, "%d", possible_count);
    *part2_result = strdup("Final star");
    
    free(lines);
}

int main() {
    FILE* file = fopen("../data/input.txt", "r");
    if (!file) {
        fprintf(stderr, "Error opening file\n");
        return 1;
    }
    
    fseek(file, 0, SEEK_END);
    long size = ftell(file);
    fseek(file, 0, SEEK_SET);
    
    char* content = (char*)malloc(size + 1);
    fread(content, 1, size, file);
    content[size] = '\0';
    fclose(file);
    
    char* part1 = NULL;
    char* part2 = NULL;
    solve(content, &part1, &part2);
    
    printf("Part 1: %s\n", part1);
    printf("Part 2: %s\n", part2);
    
    free(content);
    free(part1);
    free(part2);
    
    return 0;
}

File: days/day_12/solutions/solution.java
import java.io.*;
import java.nio.file.*;

class Solution {
    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get("../data/input.txt");
        String content = new String(Files.readAllBytes(inputPath));
        String[] results = solve(content);
        System.out.println("Part 1: " + results[0]);
        System.out.println("Part 2: " + results[1]);
    }
    
    private static String[] solve(String inputData) {
        String[] lines = inputData.trim().split("\n");
        if (lines.length == 0) {
            return new String[]{"0", "Final star"};
        }
        
        // Parse shapes (first 6 shapes, numbered 0-5)
        int[] shapeAreas = new int[6];
        int i = 0;
        int shapeIdx = 0;
        
        while (i < lines.length && shapeIdx < 6) {
            String line = lines[i].trim();
            // Check if this is a shape header (format: "number:")
            if (line.length() > 0 && line.endsWith(":")) {
                try {
                    int shapeNum = Integer.parseInt(line.substring(0, line.length() - 1));
                    if (shapeNum == shapeIdx) {
                        // Read the next 3 lines for the shape grid
                        String[] shapeGrid = new String[3];
                        for (int j = 0; j < 3; j++) {
                            if (i + 1 + j < lines.length) {
                                shapeGrid[j] = lines[i + 1 + j].trim();
                            } else {
                                shapeGrid[j] = "";
                            }
                        }
                        
                        // Count '#' characters in the shape
                        int area = 0;
                        for (String row : shapeGrid) {
                            for (char c : row.toCharArray()) {
                                if (c == '#') {
                                    area++;
                                }
                            }
                        }
                        shapeAreas[shapeIdx] = area;
                        shapeIdx++;
                        i += 4; // Skip shape header + 3 grid lines + empty line (if present)
                        continue;
                    }
                } catch (NumberFormatException e) {
                    // Not a valid shape number
                }
            }
            i++;
        }
        
        // Find where queries start (skip empty lines after shapes)
        int queryStart = i;
        while (queryStart < lines.length && lines[queryStart].trim().isEmpty()) {
            queryStart++;
        }
        
        // Parse queries
        int possibleCount = 0;
        for (int lineIdx = queryStart; lineIdx < lines.length; lineIdx++) {
            String line = lines[lineIdx].trim();
            if (line.isEmpty()) {
                continue;
            }
            
            // Parse query: "widthxheight: count0 count1 count2 count3 count4 count5"
            if (!line.contains(":")) {
                continue;
            }
            
            String[] parts = line.split(":", 2);
            if (parts.length != 2) {
                continue;
            }
            
            // Parse dimensions
            String dims = parts[0].trim();
            if (!dims.contains("x")) {
                continue;
            }
            
            String[] dimParts = dims.split("x");
            if (dimParts.length != 2) {
                continue;
            }
            
            int width, height;
            try {
                width = Integer.parseInt(dimParts[0]);
                height = Integer.parseInt(dimParts[1]);
            } catch (NumberFormatException e) {
                continue;
            }
            
            // Parse counts
            String[] countParts = parts[1].trim().split("\\s+");
            if (countParts.length != 6) {
                continue;
            }
            
            int[] counts = new int[6];
            boolean validCounts = true;
            for (int j = 0; j < 6; j++) {
                try {
                    counts[j] = Integer.parseInt(countParts[j]);
                } catch (NumberFormatException e) {
                    validCounts = false;
                    break;
                }
            }
            
            if (!validCounts) {
                continue;
            }
            
            // Calculate area check
            long regionArea = (long)width * (long)height;
            long requiredArea = 0;
            for (int j = 0; j < 6; j++) {
                requiredArea += (long)shapeAreas[j] * (long)counts[j];
            }
            
            if (requiredArea <= regionArea) {
                possibleCount++;
            }
        }
        
        // Part 2: Final star (no computation needed)
        String part2 = "Final star";
        
        return new String[]{String.valueOf(possibleCount), part2};
    }
}
